[SPARK-10688][ML][PySpark] Python API for AFTSurvivalRegression

vectorijk opened the pull request:
Implement Python API for AFTSurvivalRegression

vectorijk committed:
[SPARK-10688][ML][PySpark] Python API for AFTSurvivalRegression

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43064 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43064/consoleFull) for   PR 8926 at commit [`fa611d9`](https://github.com/apache/spark/commit/fa611d9f0634c1029e2dce93a16a1faed0651333).

SparkQA commented:
[Test build #43064 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43064/console) for   PR 8926 at commit [`fa611d9`](https://github.com/apache/spark/commit/fa611d9f0634c1029e2dce93a16a1faed0651333).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class AFTSurvivalRegressionModel(JavaModel):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43064/
Test FAILed.

mengxr commented:
@GayathriMurali Could you post comments inline? You can select the "Files changed" tab and make inline comments.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,
+                            HasFitIntercept, HasMaxIter, HasTol):
+    """
+    `https://en.wikipedia.org/wiki/Accelerated_failure_time_model`
```
It would be nice to put the link in "see also" or embed it in the doc. The first sentence is usually a summary of the class.

(@GayathriMurali  This is an inline comment.)

GayathriMurali commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+            Param(self, "quantileProbabilities",
+                  "quantile probabilities array" +
+                  ", array is not empty and every probability is in range [0,1]")
+        self.quantilesCol = Param(self, "quantilesCol",
```
Add comment to indicate what quantilesCol is

GayathriMurali commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+                  ", array is not empty and every probability is in range [0,1]")
+        self.quantilesCol = Param(self, "quantilesCol",
+                                  "quantiles column name")
+        self._setDefault(censorCol="censor",
```
Add comment to indicate what quantile probabilities is and where it gets its value from -"This column will output quantiles of corresponding quantileProbabilities if it is set."

GayathriMurali commented:
@mengxr Thanks for the input. I did add couple of inline comments, but instead of using file changed tab, i double clicked on the commit and added it. Made changes now

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+    @keyword_only
+    def __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
```
We need to use `None` for quantileProbabilities. Otherwise, the default value is mutable and shared across all `AFTSurvivalRegression` instances. See https://issues.apache.org/jira/browse/SPARK-8678 and its associated PR for details.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
+                 quantilesCol=None):
+        """
+        __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
```
Append `\` at the end to generate API doc correctly. You can run `make html` under `python/` to generate HTML docs locally and verify the output.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+        """
+        __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction",
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor",
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
```
Please keep the default array here instead of `None` because this only affects the doc.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+            "org.apache.spark.ml.regression.AFTSurvivalRegression", self.uid)
+        self.censorCol = \
+            Param(self,  "censorCol",
+                  "censor column name")
```
662-664 should fit into a single line

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -608,6 +609,141 @@ class GBTRegressionModel(TreeEnsembleModels):
+                 quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99],
+                 quantilesCol=None):
+        """
+        setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction",
```
Same about `\` at the end and the issue with mutable default value.

vectorijk committed:
clean up

vectorijk commented:
@mengxr @GayathriMurali Thanks so so much for code review! 
I cleaned up the code~

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43083 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43083/consoleFull) for   PR 8926 at commit [`b07efa3`](https://github.com/apache/spark/commit/b07efa33bf50ac8eb63d356c7e976e3a46dff40d).

SparkQA commented:
[Test build #43083 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43083/console) for   PR 8926 at commit [`b07efa3`](https://github.com/apache/spark/commit/b07efa33bf50ac8eb63d356c7e976e3a46dff40d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class AFTSurvivalRegressionModel(JavaModel):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43083/
Test PASSed.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
+        """
+        __init__(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
+                 fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor", \
+                 quantileProbabilities=None, quantilesCol=None):
```
keep the default value of `quantileProbabilities` in docstring

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
+        """
+        setParams(self, featuresCol="features", labelCol="label", predictionCol="prediction", \
+                  fitIntercept=True, maxIter=100, tol=1E-6, censorCol="censor", \
+                  quantileProbabilities=None, quantilesCol=None):
```
ditto

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
+class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,
+                            HasFitIntercept, HasMaxIter, HasTol):
+    """
+    Accelerated Failure Time(AFT) Model Survival Regression
```
space before `(` and add a period to the end.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
+
+    Fit a parametric AFT survival regression model based on the Weibull distribution
+    of the survival time.
+    see also, `https://en.wikipedia.org/wiki/Accelerated_failure_time_model`
```
use Sphinx `seealso` http://sphinx-doc.org/markup/para.html#directive-seealso Please generate the HTML doc and test the link.

vectorijk replied:
okay! I changed it and tested it locally.

mengxr commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,147 @@ class GBTRegressionModel(TreeEnsembleModels):
+    quantileProbabilities = \
+        Param(Params._dummy(), "quantileProbabilities",
+              "quantile probabilities array. Values of the quantile probabilities array " +
+              "should be in the range [0, 1] and the array should be non-empty.")
```
The range should be `(0, 1)` instead of `[0, 1]`. Btw, it would be better to keep the doc in-sync with Scala. We can do that in a separate PR.

vectorijk replied:
I looked into [params.scala#L160](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/params.scala#L160). inRange() which uses inclusive be default. Should we revise the doc or code in Scala?

yanboliang replied:
@mengxr @vectorijk  I think the range should be ```[0, 1]``` which is consistent with ```suvreg``` in R.

mengxr replied:
https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/regression/AFTSurvivalRegression.scala#L282

~~~scala
q => lambda * math.exp(math.log(-math.log(1 - q)) / k)
~~~

If q = 0 or 1, the result is NA.

vectorijk replied:
changed that in doc

mengxr commented:
LGTM except some minor inline comments

vectorijk committed:
typos

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43125/consoleFull) for   PR 8926 at commit [`f76ec8a`](https://github.com/apache/spark/commit/f76ec8ae3ede3efa7446b22b307a09087f4e4012).

yanboliang commented on python/pyspark/ml/regression.py:
```diff
@@ -609,6 +611,150 @@ class GBTRegressionModel(TreeEnsembleModels):
+                  quantileProbabilities=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99], \
+                  quantilesCol=None):
+        """
+        kwargs = self.__init__._input_kwargs
```
I think it's need special handling for ```quantileProbabilities``` here.
Because that if users did not set ```quantileProbabilities```, it will be set with ```None``` which is not default value ```[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]``` in the current route.
You can refer the [```layers```](https://github.com/apache/spark/blob/master/python/pyspark/ml/classification.py#L833) of MultilayerPerceptronClassifier.

vectorijk replied:
@mengxr @yanboliang changed that.
Plus, I added two methods in AFTSurvivalRegressionModel.

SparkQA commented:
[Test build #43125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43125/console) for   PR 8926 at commit [`f76ec8a`](https://github.com/apache/spark/commit/f76ec8ae3ede3efa7446b22b307a09087f4e4012).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class AFTSurvivalRegressionModel(JavaModel):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43125/
Test PASSed.

vectorijk committed:
implement AFTSurvivalRegressionModel methods

- set quantileProbabilites by array in setParams
- two methods in AFTSurvivalRegressionModel

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43139 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43139/consoleFull) for   PR 8926 at commit [`25ac136`](https://github.com/apache/spark/commit/25ac136a80bbd37c0fbc01392831ffadb4803309).

SparkQA commented:
[Test build #43139 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43139/console) for   PR 8926 at commit [`25ac136`](https://github.com/apache/spark/commit/25ac136a80bbd37c0fbc01392831ffadb4803309).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class AFTSurvivalRegressionModel(JavaModel):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43139/
Test PASSed.

vectorijk committed:
change docs in quantileProbabilities

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43145 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43145/consoleFull) for   PR 8926 at commit [`4f1e8a9`](https://github.com/apache/spark/commit/4f1e8a9e008e251f1da50fccd313db72f3b87c18).

SparkQA commented:
[Test build #43145 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43145/console) for   PR 8926 at commit [`4f1e8a9`](https://github.com/apache/spark/commit/4f1e8a9e008e251f1da50fccd313db72f3b87c18).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class AFTSurvivalRegression(JavaEstimator, HasFeaturesCol, HasLabelCol, HasPredictionCol,`
  * `class AFTSurvivalRegressionModel(JavaModel):`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43145/
Test PASSed.

vectorijk commented:
@mengxr  Could you review again cause I have changed it all?

mengxr commented:
LGTM. Merged into master. I sent https://github.com/apache/spark/pull/9001 for a minor bug fix. Thanks!

vectorijk commented:
@mengxr  Thanks so much for reviewing! It is my first PR to spark community. Excited!
P.S As you mentioned before, range of quantile probabilities is (0,1) instead of [0,1]. Should we create a PR for changing that in corresponding scala code?

mengxr commented:
Yes. We should fix it. Could you make a JIRA and submit a PR? Thanks!

vectorijk commented:
Sure
